part-1.py

```diff
@@ -4,17 +4,41 @@
 # the appropriate comment.
 
 # factorial
-
+def factorial(n):
+    if n == 0:
+        return 1
+    elif n < 0:
+        raise ValueError
+    else:
+        return n * factorial(n - 1)
 
 
 # reverse
+def reverse(text):
+    if not text:
+        return text
+    else:
+        return text[-1] + reverse(text[0:-1])
 
 
 
 # bunny
-
+def bunny(count): 
+    if count == 0:
+        return 0
+    else:
+        return 2 + bunny(count - 1)
 
 
 # is_nested_parens
+def is_nested_parens(parens):
+    if not parens:
+        return True
+    elif len(parens) % 2 != 0:
+        return False
```

Comment:
This works, but it's going to short circuit the recursion.  A recursive base case is something more like:
```suggestion
    elif len(parens) == 1:
        return False
```